File: 00_raw_data/isolate_tree/workflow/scripts/concatenate_alignments.py
#!/usr/bin/env python3
"""
Concatenate multiple gene alignments into a single super-alignment.
Renames sequences using isolate IDs instead of genome codes.
"""

from Bio import SeqIO
from Bio.Seq import Seq
from Bio.SeqRecord import SeqRecord
import pandas as pd
from collections import defaultdict

# Read inputs from snakemake
alignment_files = snakemake.input["alignments"]
mapping_file = snakemake.input["mapping"]
output_file = snakemake.output[0]

# Read the genome code to isolate ID mapping
mapping_df = pd.read_csv(mapping_file, sep="\t")
code_to_isolate = dict(zip(mapping_df["genome_code"], mapping_df["isolate_id"]))

# Store sequences for each genome across all gene alignments
# Structure: {genome_code: {gene_file: sequence}}
genome_sequences = defaultdict(dict)
gene_lengths = {}  # Track alignment length for each gene

# Read all alignment files
for aln_file in alignment_files:
    gene_name = aln_file  # Store the file path as identifier
    sequences = SeqIO.parse(aln_file, "fasta")

    # Track alignment length (should be same for all sequences in this file)
    aln_length = None

    for record in sequences:
        # Extract genome code (first 6 characters of sequence ID)
        genome_code = str(record.id)[:6]

        # Store the sequence
        genome_sequences[genome_code][gene_name] = str(record.seq)

        # Record alignment length
        if aln_length is None:
            aln_length = len(record.seq)

    gene_lengths[gene_name] = aln_length

# Get all unique genome codes
all_genome_codes = sorted(genome_sequences.keys())

# Build concatenated sequences
concatenated_records = []

for genome_code in all_genome_codes:
    concat_seq = ""

    # For each gene alignment, add the sequence or gaps if missing
    for aln_file in alignment_files:
        if aln_file in genome_sequences[genome_code]:
            concat_seq += genome_sequences[genome_code][aln_file]
        else:
            # Genome is missing this gene - add gaps
            print(aln_file, genome_code)
            print(gene_lengths)
            concat_seq += "-" * gene_lengths[aln_file]

    # Map genome code to isolate ID
    isolate_id = code_to_isolate.get(genome_code, genome_code)

    # Create SeqRecord
    record = SeqRecord(
        Seq(concat_seq),
        id=isolate_id,
        description=""
    )
    concatenated_records.append(record)

# Write concatenated alignment
SeqIO.write(concatenated_records, output_file, "fasta")

print(f"Concatenated {len(alignment_files)} gene alignments")
print(f"Total alignment length: {len(concatenated_records[0].seq)} bp")
print(f"Number of genomes: {len(concatenated_records)}")

File: 00_raw_data/isolate_tree/workflow/scripts/concatenate_alignments.R
library(tidyverse)

# Read the genome code mapping
mapping <- read_tsv(snakemake@input[["mapping"]], col_types = "cc")

# Get list of alignment files
alignment_files <- snakemake@input[["alignments"]]

# Function to read a FASTA file
read_fasta <- function(file) {
  lines <- readLines(file)
  headers <- lines[grepl("^>", lines)]
  sequences <- lines[!grepl("^>", lines)]

  # Group sequences by header
  header_indices <- which(grepl("^>", lines))

  result <- list()
  for (i in seq_along(header_indices)) {
    header <- lines[header_indices[i]]
    # Get sequence lines until next header or end
    if (i < length(header_indices)) {
      seq_lines <- lines[(header_indices[i] + 1):(header_indices[i + 1] - 1)]
    } else {
      seq_lines <- lines[(header_indices[i] + 1):length(lines)]
    }
    sequence <- paste(seq_lines, collapse = "")

    # Extract genome code (first 6 characters after >)
    genome_code <- str_sub(str_remove(header, "^>"), 1, 6)

    result[[genome_code]] <- sequence
  }

  return(result)
}

# Read all alignments
all_alignments <- list()
for (aln_file in alignment_files) {
  gene_seqs <- read_fasta(aln_file)
  all_alignments[[basename(aln_file)]] <- gene_seqs
}

# Get all unique genome codes across all alignments
all_genome_codes <- unique(unlist(lapply(all_alignments, names)))

# Build concatenated alignment for each genome
concatenated <- list()
for (genome_code in all_genome_codes) {
  concat_seq <- ""
  for (aln_name in names(all_alignments)) {
    gene_seqs <- all_alignments[[aln_name]]
    if (genome_code %in% names(gene_seqs)) {
      concat_seq <- paste0(concat_seq, gene_seqs[[genome_code]])
    } else {
      # If genome missing this gene, add gaps
      # Get length from first available sequence in this alignment
      gap_length <- nchar(gene_seqs[[1]])
      concat_seq <- paste0(concat_seq, paste(rep("-", gap_length), collapse = ""))
    }
  }
  concatenated[[genome_code]] <- concat_seq
}

# Map genome codes to isolate IDs
mapping_lookup <- setNames(mapping$isolate_id, mapping$genome_code)

# Write output FASTA with isolate IDs as headers
output_file <- snakemake@output[[1]]
writeLines("", output_file)  # Create/clear file

for (genome_code in names(concatenated)) {
  isolate_id <- mapping_lookup[genome_code]
  if (is.na(isolate_id)) {
    isolate_id <- genome_code  # Fallback if not in mapping
  }

  # Write header
  cat(paste0(">", isolate_id, "\n"), file = output_file, append = TRUE)

  # Write sequence (wrap at 80 characters for readability)
  sequence <- concatenated[[genome_code]]
  seq_wrapped <- gsub("(.{80})", "\\1\n", sequence)
  cat(seq_wrapped, "\n", file = output_file, append = TRUE, sep = "")
}
